### PR TITLE
✨ Improve token permission check

### PR DESCRIPTION
**Pull request by laurentsimon**
* **Please check if the PR fulfills these requirements**
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] PR title follows the guidelines defined in  https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
feature

* **What is the current behavior?** (You can also link to an open issue here)
we only support top-level permissions

* **What is the new behavior (if this is a feature change)?**
we support run-level permissions

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
no

* **Other information**:
1. We don't flag the `packages` permission if the workflow is a packaging workflow. This is good enough for now. Ideally we should only check for packaging workflows that publish to GitHub registries (since the `packages` permission is only needed in this case).
2. We don't check for `security-events` permission if the workflow is codeql. We may add other SAST tool in the future
3. I added support to the Packaging check for some missing registries/language like go, ruby's gem and nuget.

### CODE REVIEW

File: checks/packaging.go [hunk 1]
```diff
@@ -77,49 +77,75 @@ func Packaging(c *checker.CheckRequest) checker.CheckResult {
...
+
+	// Python packages.
+	if strings.Contains(s, "actions/setup-python@") && strings.Contains(s, "pypa/gh-action-pypi-publish@master") {
+		dl.Info("candidate python packaging workflow using pypi: %s", fp)
 		return true
 	}
 
-	if strings.Contains(s, "docker push") {
-		c.Dlogger.Info("candidate docker publishing workflow: %s", fp)
+	// Go packages.
```
Comment (oliverchang) [inline checks/packaging.go:142]: Let's add support for Rust/Cargo as well?

File: checks/permissions.go [hunk 2]
```diff
@@ -72,41 +78,46 @@ func validatePermission(key string, value interface{}, path string,
...
+	ignoredPermissions map[string]bool,
+	acceptUndefined bool) error {
 	var permissions interface{}
 
 	// Check if permissions are set explicitly.
 	permissions, ok := config["permissions"]
 	if !ok {
-		dl.Warn("no permission defined in %v", path)
-		recordAllPermissionsWrite(pdata)
+		if !acceptUndefined {
```
Comment (oliverchang) [inline checks/permissions.go:117]: would be helpful to add a short comment here to explain why there's this condition. e.g. "Explicit permissions are optional for run level" or something similar. 

Also might be clearer to move the `permissions, ok := config["permissions"]` bits and this check to where they're called from (i.e. validateRunLevelPermissions / validateTopLevelPermissions) and pass the resulting `permissions` to this function instead. This makes the special case (Warn + recordAllPermissionsWrite) closer to where it's needed and easier to follow.

i.e. 

```go
func validateTopLevelPermissions(config map[interface{}]interface{}, path string,
	dl checker.DetailLogger, pdata *permissionCbData) error {
        permissions, ok := config["permissions"]
        // comment explaining why.
        if !ok {
             ... // warn + recordAllPermissionsWrite
            return
        }
	return validatePermissions(permissions, path, dl,
		pdata.topLevelWritePermissions, map[string]bool{},
		false)
}
```

**Review by oliverchang — APPROVED** [review]

File: checks/packaging.go (hunk 1 above)
Comment (laurentsimon) [reply checks/packaging.go:142]: done.

File: checks/permissions.go (hunk 2 above)
Comment (laurentsimon) [reply checks/permissions.go]: good idea. I started adding more and more arguments without realizing it could be simplified. Thanks!

**Pull request merged**